### DISCUSSION TITLE
Obtaining Syntax Mark Information from MDAST

### DISCUSSION THREAD

**Discussion by eliyap** [micromark/micromark#88] [Ideas] [answered]
Hi All!

Is it possible to get syntax marks positions / ranges from MDAST?

For `**strong**` or `_emph_`, it's straightforwards.

Fenced code is trickier. They can open with 3+ backticks or tildes, and the ending fence may be longer than the opening fence.
AFAIK, MDAST exposes `Code`'s *overall* range, `lang` and `meta` attributes, but not opening / closing range. 

It's possible to search for `value` within the `Literal` range, but official support help in corner cases:
- \`\`\`\`\`\` (an empty fence with 3 open + 3 close backticks)
- \`\`\`\`\`\`\`\` (3 open + 5 close? 4 open + 4 close?)

Lastly, I'd love this for Blockquotes, which are most difficult. `>` are interleaved within contents, may / may not be present (laziness clause), may / may not have a space after them, and may / may not be nested.

Appreciate any thoughts!

**Comment by wooorm** [comment, marked as answer]
What you’re looking for is called a CST. See also: https://github.com/micromark/micromark/issues/59.

micromark allows creating one, and was made for that, but one hasn’t been created yet.

*Why* do you want this?

**Reply by eliyap** [reply]
Sorry for the late response 

Thank you for the clarification, a CST is exactly what I’m looking for! 
My motivation is to improve “smart syntax”, see https://pencil.md (previously discussed [here](https://github.com/micromark/micromark/discussions/85)) 

Take the following
```markdown
I met a traveller from an antique land, who said:
> I met a traveller from an antique land, who said:
> > I met a traveller from an antique land, who said: ...
> > et cetera
> > et cetera
```
Say the user (in my app) strikes out / erases the entire last line – including the `>`'s! 
In this case, I consider them to be targeting the text, not the syntax marks. (though we might clean up the syntax marks if the block is empty)

So we would wish to obtain:
```markdown
I met a traveller from an antique land, who said:
> I met a traveller from an antique land, who said:
> > I met a traveller from an antique land, who said: ...
> > et cetera
> > ~~et cetera~~
```
instead of
```markdown
I met a traveller from an antique land, who said:
> I met a traveller from an antique land, who said:
> > I met a traveller from an antique land, who said: ...
> > et cetera
~~> > et cetera~~
```

AFAIK, the MDAST can't tell me which `>` belongs to which blockquote, and due to the [laziness clause](https://github.github.com/gfm/#example-210), an educated guess is tricky. 

I understand that CSTs aren't in high demand and might take a lot of work, so it it's not on the table that's ok too!

**Reply by wooorm** [reply]
laziness: yeah, tricky indeed!

CST: It’s not really on the todolist for me to work on unfortunately, I have a ton of other things to do with a higher priority.

Best!
